optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/DroolsConstraintSessionFactory.java

```diff
@@ -103,41 +98,93 @@ public boolean isDroolsAlphaNetworkCompilationEnabled() {
-            scoreHolder.configureConstraintWeight(compiledRule, constraintWeight);
-            if (constraintWeight.equals(zeroScore)) {
-                disabledConstraintIdSet.add(constraint.getConstraintId());
-            }
-        });
-        // Determine the KieBase to use.
-        if (disabledConstraintIdSet.isEmpty()) { // Shortcut; don't change the original KieBase.
-            currentKieBase = originalKieBase;
-            currentlyDisabledConstraintIdSet = null;
-        } else if (!disabledConstraintIdSet.equals(currentlyDisabledConstraintIdSet)) {
-            // Only rebuild the active KieBase when the set of disabled constraints changed.
-            ModelImpl model = new ModelImpl().withGlobals(originalModel.getGlobals());
-            constraintToModelRuleMap.forEach((constraintId, modelRule) -> {
-                if (disabledConstraintIdSet.contains(constraintId)) {
-                    return;
-                }
-                model.addRule(modelRule);
-            });
-            currentKieBase = buildKieBaseFromModel(model, droolsAlphaNetworkCompilationEnabled);
-            currentlyDisabledConstraintIdSet = disabledConstraintIdSet;
+
+        // Extract constraint weights, excluding constraints where weight is zero.
+        Map<DroolsConstraint<Solution_>, Score_> constraintToWeightMap =
+                InnerConstraintFactory.extractConstraintToWeightMap(constraintList,
+                        c -> (Score_) c.extractConstraintWeight(workingSolution), zeroScore);
+
+        // Creating KieBase is expensive. Therefore we only do it when there has been a change in constraint weights.
+        if (kieBaseCache == null || !kieBaseCache.isUpToDate(constraintToWeightMap)) {
+            Package pack = solutionDescriptor.getSolutionClass().getPackage();
+            String constraintPackageName = (pack == null) ? "" : pack.getName();
+            // Each constraint gets its own global, in which it will keep its impacter.
+            // Impacters carry constraint weights, and therefore the instances are solution-specific.
+            AtomicInteger idCounter = new AtomicInteger(0);
+            Map<DroolsConstraint<Solution_>, Global<WeightedScoreImpacter>> constraintToGlobalMap =
+                    constraintToWeightMap.keySet().stream()
+                            .collect(Collectors.toMap(Function.identity(),
+                                    c -> globalOf(WeightedScoreImpacter.class, constraintPackageName,
+                                            "scoreImpacter" + idCounter.getAndIncrement())));
+            ModelImpl model = constraintToWeightMap.keySet().stream()
+                    .map(constraint -> constraint.buildRule(constraintToGlobalMap.get(constraint)))
+                    .reduce(new ModelImpl(), ModelImpl::addRule, (m, __) -> m);
+            constraintToGlobalMap.forEach((constraint, global) -> model.addGlobal(global));
```

Comment:
Instead of 1 ScoreHolder global per session, we now have a ScoreImpacter global per constraint per session.
This is necessary to guarantee that constraint are independent across sessions, and that constraint weights are solution-specific.

This has no impact on performance.

Reply:
"we now have a ScoreImpacter global per constraint per session". What if there are 100 constraints? How - and when - is the score summed up?

Reply:
The same way it always was - there is only one ScoreInliner.